runtime/near-vm-logic/src/logic.rs

```diff
@@ -2786,6 +2786,67 @@ impl<'a> VMLogic<'a> {
         }
         Ok(())
     }
+
+    /// Verify an ED25519 signature given a message and a public key.
+    /// # Cost
+    ///
+    /// `base + write_register_base + write_register_byte * num_bytes + ed25519_verify_base + ed25519_verify_byte * num_bytes`
+
+    pub fn ed25519_verify(
+        &mut self,
+        sig_len: u64,
+        sig_ptr: u64,
+        msg_len: u64,
+        msg_ptr: u64,
+        pub_key_len: u64,
+        pub_key_ptr: u64,
+        register_id: u64,
+    ) -> Result<()> {
+        use ed25519_dalek::{PublicKey, Signature, Verifier};
+
+        self.gas_counter.pay_base(ed25519_verify_base)?;
+        self.gas_counter.pay_per(ed25519_verify_byte, msg_len as u64)?;
+        let signature_array = self.get_vec_from_memory_or_register(sig_ptr, sig_len)?;
+        let msg = self.get_vec_from_memory_or_register(msg_ptr, msg_len)?;
+        let pub_key_array = self.get_vec_from_memory_or_register(pub_key_ptr, pub_key_len)?;
+        let pub_key = PublicKey::from_bytes(&pub_key_array)
+            .map_err(|e| VMLogicError::ExternalError(AnyError::new(e.to_string())))?;
+        let signature = Signature::from_bytes(&signature_array)
+            .map_err(|e| VMLogicError::ExternalError(AnyError::new(e.to_string())))?;
+
+        let value = pub_key.verify(&msg, &signature).is_ok();
+        self.internal_write_register(register_id, vec![u8::from(value)])
+    }
+
+    /// Verify an Schnorrkel signature given a message and a public key.
+    /// # Cost
+    ///
+    /// `base + write_register_base + write_register_byte * num_bytes + sr25519_verify_base + sr25519_verify_byte * num_bytes`
+    pub fn sr25519_verify(
+        &mut self,
+        sig_len: u64,
+        sig_ptr: u64,
+        msg_len: u64,
+        msg_ptr: u64,
+        pub_key_len: u64,
+        pub_key_ptr: u64,
+        register_id: u64,
+    ) -> Result<()> {
+        use schnorrkel::{PublicKey, Signature};
+
+        self.gas_counter.pay_base(sr25519_verify_base)?;
+        self.gas_counter.pay_per(sr25519_verify_byte, msg_len as u64)?;
+        let signature_array = self.get_vec_from_memory_or_register(sig_ptr, sig_len)?;
+        let msg = self.get_vec_from_memory_or_register(msg_ptr, msg_len)?;
+        let pub_key_array = self.get_vec_from_memory_or_register(pub_key_ptr, pub_key_len)?;
+        let pub_key = PublicKey::from_bytes(&pub_key_array)
+            .map_err(|e| VMLogicError::ExternalError(AnyError::new(e.to_string())))?;
+        let signature = Signature::from_bytes(&signature_array)
+            .map_err(|e| VMLogicError::ExternalError(AnyError::new(e.to_string())))?;
+
+        let value = pub_key.verify_simple(b"substrate", &msg, &signature).is_ok();
```

Comment:
This `substrate` constant feels suspicious, why is it here? I haven't familiarized myself with the context here, so there might be perfect sense for it, but it warrants at least a comment.

Reply:
this is old, sorry! I'll clean it and ask for a new review

Reply:
context: this PR had multiple host functions. Now we've only are adding signature verification on ED25519